hieroplatform/src/main/java/org/hiero/sketch/spreadsheet/HeatMap.java

```diff
@@ -0,0 +1,132 @@
+        final IRowIterator myIter = membershipSet.getIterator();
+        int currRow = myIter.getNextRow();
+        while (currRow >= 0) {
+            if ((columnD1.isMissing(currRow)) || (columnD2.isMissing(currRow))){
+                if (!columnD1.isMissing(currRow))  //only column 2 is missing
+                    this.histogramMissingD1.addValue(columnD1.asDouble(currRow, converterD1));
+                else if (!columnD2.isMissing(currRow)) // only column 1 is missing
+                    this.histogramMissingD2.addValue(columnD2.asDouble(currRow, converterD2));
+                else
+                    this.missingData++; // both are missing
+                }
+            else {
+                double val1 = columnD1.asDouble(currRow,converterD1);
+                double val2 = columnD2.asDouble(currRow,converterD2);
+                int index1 = this.bucketDescDim1.indexOf(val1);
+                int index2 = this.bucketDescDim2.indexOf(val2);
+                if ((index1 >= 0) && (index2 >= 0)) {
+                    this.buckets[index1][index2]++;
+                    this.totalsize++;
+                }
+                else this.outOfRange++;
+            }
+            currRow = myIter.getNextRow();
+        }
+    }
+
+    public Histogram1DLight getMissingHistogramD1() { return this.histogramMissingD1; }
+
+    public long getSize() { return this.totalsize; }
+
+
+    public Histogram1DLight getMissingHistogramD2() { return this.histogramMissingD2; }
+
+    public void createSampleHistogram(final IColumn columnD1, final IColumn columnD2,
+                                      final IStringConverter converterD1, final IStringConverter converterD2,
+                                      final IMembershipSet membershipSet, double sampleRate) {
+        this.createHistogram(columnD1, columnD2, converterD1, converterD2, membershipSet.sample(sampleRate));
+    }
+
+    public void createSampleHistogram(final IColumn columnD1, final IColumn columnD2,
+                                      final IStringConverter converterD1, final IStringConverter converterD2,
+                                      final IMembershipSet membershipSet, double sampleRate, long seed) {
+        this.createHistogram(columnD1, columnD2, converterD1, converterD2, membershipSet.sample(sampleRate, seed));
+    }
+
+    public int getNumOfBucketsD1() { return this.bucketDescDim1.getNumOfBuckets(); }
+
+    public int getNumOfBucketsD2() { return this.bucketDescDim2.getNumOfBuckets(); }
+
+    public long getMissingData() { return this.missingData; }
+
+    public long getOutOfRange() { return this.outOfRange; }
+
+    /**
+     * @return the index's bucket or null if not been initialized yet
+     */
+    public long getBucket(final int index1, final int index2) {
+        if (!this.initialized)
+            throw new IllegalArgumentException("bucket not initialized yet");
+        if ((index1 < 0) || (index1 >= this.bucketDescDim1.getNumOfBuckets())
```

Comment:
The bounds checks are redundant: the access below to this.buckets will do them anyway.
I am not sure that the "getBucket" name is appropriate. Maybe getCount?